core/PruManager.cpp

```diff
@@ -0,0 +1,156 @@
+/*
+ * PruManager.cpp
+ *
+ *	Created on: July 3, 2021
+ *		Author: Dhruva Gole
+ */
+
+#include "../include/PruManager.h"
+#include "MiscUtilities.h"
+#include <iostream>
+
+#if ENABLE_PRU_RPROC == 1
+PruManagerRprocMmap::PruManagerRprocMmap(unsigned int pruNum, unsigned int v)
+{
+	/* constructor for initializing the necessary path variables
+	 * based on the value of pru_num to choose:
+	 * 0 for PRU1 core 0
+	 * 1 for PRU1 core 1
+	 * 2 for PRU2 core 0
+	 * 3 for PRU2 core 1
+	 */
+	pru_num = pruNum;
+	verbose = v;
+	unsigned int pruss = pru_num / 2 + 1;
+	unsigned int prucore = pru_num % 2;
+
+	basePath = "/dev/remoteproc/pruss" + std::to_string(pruss) + "-core" + std::to_string(prucore) + "/";
+	statePath = basePath + "state";
+	firmwarePath = basePath + "firmware";
+	firmware = "am57xx-pru" + std::to_string(pruss) + "_" + std::to_string(prucore) + "-fw";
+# ifdef firmwareBelaRProc
+	std::string firmwareBela = firmwareBelaRProc;	// Incoming string from Makefile
+# endif	// firmwareBelaRProc
+	firmwareCopyCommand = "sudo ln -s -f " + firmwareBela + " /lib/firmware/" + firmware; // Pass the name of custom .out file via Makefile using firmwareBelaRProc=<path to file>
+	// 0 : pru1-core 0 in AI -> 4b234000
+	// 1 : pru1-core 1 in AI -> 4b238000
+	// 2 : pru2-core 0 in AI -> 4b2b4000
+	// 3 : pru2-core 1 in AI -> 4b2b8000
+	//
+	// 1 : PRUSS address in AI for PRU 1
+	// 2 : PRUSS address in AI for PRU 2
+	//
+	// 0 : pru-core 0 in BBB -> 4a334000
+	// 1 : pru-core 1 in BBB -> 4a338000
+
+# ifdef IS_AM572x	// base addresses for BBAI
+	pru_addr.insert(std::pair<unsigned int, unsigned int>(0,0x4b234000));
+	pru_addr.insert(std::pair<unsigned int, unsigned int>(1,0x4b238000));
+	pru_addr.insert(std::pair<unsigned int, unsigned int>(2,0x4b2b4000));
+	pru_addr.insert(std::pair<unsigned int, unsigned int>(3,0x4b2b8000));
+
+	pruss_addr.insert(std::pair<unsigned int, unsigned int>(1,0x4b200000));
+	pruss_addr.insert(std::pair<unsigned int, unsigned int>(2,0x4b280000));
+# else	// base addresses for BBB
+	pru_addr.insert(std::pair<unsigned int, unsigned int>(0,0x4a334000));
+	pru_addr.insert(std::pair<unsigned int, unsigned int>(1,0x4a338000));
+# endif	// IS_AM572x
+}
+
+void PruManagerRprocMmap::stop()
+{	// performs echo stop > state
+	if(verbose)
+		std::cout << "Stopping the PRU" << std::to_string(pruss) + "_" + std::to_string(prucore) << "\n";
+	IoUtils::writeTextFile(statePath, "stop");
+}
+
+int PruManagerRprocMmap::start(const std::string& path)
+{
+	stop();
+	system(firmwareCopyCommand.c_str());	// copies fw to /lib/am57xx-fw
+	if(verbose)
+		std::cout << "Loading firmware into the PRU" << std::to_string(pruss) + "_" + std::to_string(prucore) << "\n";
```

Comment:
in this case for instance you clearly see the sheer power and elegance of `printf()` against `std::cout`.